File: scripts/metrics-agent.py
#!/usr/bin/env python3
"""
Metrics Agent for GPU Nodes (Chimera/Cerberus)
Serves system and GPU metrics as JSON on port 9100
"""

import json
import subprocess
import os
import psutil
from http.server import HTTPServer, BaseHTTPRequestHandler
from datetime import datetime

PORT = int(os.environ.get('METRICS_PORT', 9100))

def get_gpu_metrics():
    """Get GPU metrics using nvidia-smi"""
    gpus = []
    try:
        result = subprocess.run(
            ['nvidia-smi', '--query-gpu=index,name,utilization.gpu,memory.used,memory.total,temperature.gpu',
             '--format=csv,noheader,nounits'],
            capture_output=True, text=True, timeout=10
        )
        if result.returncode == 0:
            for line in result.stdout.strip().split('\n'):
                if line:
                    parts = [p.strip() for p in line.split(',')]
                    if len(parts) >= 6:
                        gpus.append({
                            'index': int(parts[0]),
                            'name': parts[1],
                            'utilization_percent': int(parts[2]) if parts[2] != '[N/A]' else 0,
                            'memory_used_gb': round(int(parts[3]) / 1024, 1) if parts[3] != '[N/A]' else 0,
                            'memory_total_gb': round(int(parts[4]) / 1024, 1) if parts[4] != '[N/A]' else 0,
                            'temperature_c': int(parts[5]) if parts[5] != '[N/A]' else 0
                        })
    except Exception as e:
        print(f"Error getting GPU metrics: {e}")
    return gpus

def get_system_metrics():
    """Get CPU, RAM, and disk metrics"""
    # CPU
    cpu_percent = psutil.cpu_percent(interval=0.5)

    # Memory
    mem = psutil.virtual_memory()
    ram_used_gb = round(mem.used / (1024**3), 1)
    ram_total_gb = round(mem.total / (1024**3), 1)

    # Disk - root partition
    disk = psutil.disk_usage('/')
    disk_used_gb = round(disk.used / (1024**3), 1)
    disk_total_gb = round(disk.total / (1024**3), 1)

    return {
        'cpu_percent': round(cpu_percent),
        'ram_used_gb': ram_used_gb,
        'ram_total_gb': ram_total_gb,
        'disk_used_gb': disk_used_gb,
        'disk_total_gb': disk_total_gb
    }

def get_container_count():
    """Get count of running Docker containers"""
    try:
        result = subprocess.run(
            ['docker', 'ps', '-q'],
            capture_output=True, text=True, timeout=10
        )
        if result.returncode == 0:
            return len([l for l in result.stdout.strip().split('\n') if l])
    except:
        pass
    return 0

def collect_metrics():
    """Collect all metrics"""
    return {
        'timestamp': datetime.utcnow().isoformat() + 'Z',
        'system': get_system_metrics(),
        'gpus': get_gpu_metrics(),
        'containers': {
            'running': get_container_count()
        }
    }

class MetricsHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        # Suppress default logging
        pass

    def do_GET(self):
        if self.path == '/metrics' or self.path == '/':
            metrics = collect_metrics()
            response = json.dumps(metrics, indent=2)

            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(response.encode())
        elif self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(b'OK')
        else:
            self.send_response(404)
            self.end_headers()

def main():
    server = HTTPServer(('0.0.0.0', PORT), MetricsHandler)
    print(f"Metrics agent listening on port {PORT}")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        server.shutdown()

if __name__ == '__main__':
    main()

File: agents/metrics-agent.js
#!/usr/bin/env node
// Hydra Cluster Metrics Agent
// Collects real system and GPU metrics for cluster monitoring
// Deploy on Chimera, Cerberus, or any GPU node

const http = require('http');
const { execSync } = require('child_process');
const os = require('os');

const NODE_NAME = process.env.NODE_NAME || os.hostname();
const NODE_ROLE = process.env.NODE_ROLE || 'worker';
const PORT = parseInt(process.env.PORT || '9100');

// Cache for GPU metrics (nvidia-smi can be slow)
let gpuCache = { data: [], timestamp: 0 };
const GPU_CACHE_TTL = 5000; // 5 seconds

function getGpuMetrics() {
  const now = Date.now();
  if (now - gpuCache.timestamp < GPU_CACHE_TTL) {
    return gpuCache.data;
  }

  const gpus = [];
  try {
    // Query nvidia-smi for GPU info
    const output = execSync(
      'nvidia-smi --query-gpu=index,name,utilization.gpu,memory.used,memory.total,temperature.gpu,power.draw --format=csv,noheader,nounits',
      { encoding: 'utf8', timeout: 10000 }
    );

    const lines = output.trim().split('\n');
    for (const line of lines) {
      const parts = line.split(',').map(s => s.trim());
      if (parts.length >= 6) {
        gpus.push({
          index: parseInt(parts[0]) || 0,
          name: parts[1] || 'Unknown GPU',
          utilization_percent: parseInt(parts[2]) || 0,
          memory_used_gb: Math.round((parseInt(parts[3]) || 0) / 1024 * 10) / 10,
          memory_total_gb: Math.round((parseInt(parts[4]) || 0) / 1024 * 10) / 10,
          temperature_c: parseInt(parts[5]) || 0,
          power_draw_w: Math.round(parseFloat(parts[6]) || 0)
        });
      }
    }
  } catch (e) {
    console.error('[metrics-agent] nvidia-smi failed:', e.message);
  }

  gpuCache = { data: gpus, timestamp: now };
  return gpus;
}

function getSystemMetrics() {
  // CPU usage from /proc/stat
  let cpuPercent = 0;
  try {
    const output = execSync(
      "top -bn1 | grep 'Cpu(s)' | awk '{print $2}'",
      { encoding: 'utf8', timeout: 5000 }
    );
    cpuPercent = Math.round(parseFloat(output.trim()) || 0);
  } catch (e) {
    // Fallback to node.js calculation
    const cpus = os.cpus();
    cpuPercent = Math.round(cpus.reduce((acc, cpu) => {
      const total = Object.values(cpu.times).reduce((a, b) => a + b, 0);
      const idle = cpu.times.idle;
      return acc + ((total - idle) / total) * 100;
    }, 0) / cpus.length);
  }

  // Memory
  const totalMem = os.totalmem();
  const freeMem = os.freemem();
  const usedMem = totalMem - freeMem;

  // Disk usage
  let diskUsedGb = 0;
  let diskTotalGb = 1000;
  try {
    const dfOutput = execSync("df -BG / | tail -1 | awk '{print $2, $3}'", { encoding: 'utf8', timeout: 5000 });
    const [total, used] = dfOutput.trim().split(/\s+/).map(s => parseInt(s.replace('G', '')));
    if (!isNaN(total)) diskTotalGb = total;
    if (!isNaN(used)) diskUsedGb = used;
  } catch (e) {
    // Use defaults
  }

  // Load average
  const loadAvg = os.loadavg();

  // Uptime
  const uptimeHours = Math.round(os.uptime() / 3600);

  return {
    cpu_percent: cpuPercent,
    ram_used_gb: Math.round(usedMem / (1024 * 1024 * 1024)),
    ram_total_gb: Math.round(totalMem / (1024 * 1024 * 1024)),
    disk_used_gb: diskUsedGb,
    disk_total_gb: diskTotalGb,
    load_average: loadAvg.map(x => Math.round(x * 100) / 100),
    uptime_hours: uptimeHours
  };
}

function getContainerMetrics() {
  let running = 0;
  let paused = 0;
  let stopped = 0;

  try {
    const output = execSync('docker ps -a --format "{{.State}}" 2>/dev/null', { encoding: 'utf8', timeout: 5000 });
    const states = output.trim().split('\n').filter(Boolean);
    for (const state of states) {
      if (state === 'running') running++;
      else if (state === 'paused') paused++;
      else stopped++;
    }
  } catch (e) {
    // Docker not available or permission denied
  }

  return { running, paused, stopped };
}

function getQueueMetrics() {
  // Placeholder - could be integrated with a job queue system
  return {
    pending: 0,
    estimated_wait_minutes: 0
  };
}

function getActiveTrainingJob() {
  // Check for common training processes
  try {
    // Look for python processes with common training framework indicators
    const output = execSync(
      "ps aux | grep -E 'python.*train|python.*fine|torchrun|deepspeed' | grep -v grep | head -1",
      { encoding: 'utf8', timeout: 5000 }
    );
    if (output.trim()) {
      // Extract process name from command line
      const cmd = output.split(/\s+/).slice(10).join(' ').substring(0, 50);
      return {
        name: cmd || 'training-job',
        progress_percent: null, // Can't determine without integration
        eta_minutes: null
      };
    }
  } catch (e) {
    // No training job found
  }
  return null;
}

const server = http.createServer((req, res) => {
  if (req.url === '/metrics' && req.method === 'GET') {
    const metrics = {
      hostname: NODE_NAME,
      role: NODE_ROLE,
      timestamp: new Date().toISOString(),
      gpus: getGpuMetrics(),
      system: getSystemMetrics(),
      containers: getContainerMetrics(),
      queue: getQueueMetrics(),
      active_training_job: getActiveTrainingJob()
    };

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(metrics, null, 2));
  } else if (req.url === '/health' && req.method === 'GET') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ status: 'healthy', node: NODE_NAME }));
  } else {
    res.writeHead(404);
    res.end('Not Found');
  }
});

server.listen(PORT, '0.0.0.0', () => {
  console.log(`[${NODE_NAME}] Metrics agent running on port ${PORT}`);
  console.log(`[${NODE_NAME}] Role: ${NODE_ROLE}`);
  console.log(`[${NODE_NAME}] Endpoints: GET /metrics, GET /health`);
});

// Graceful shutdown
process.on('SIGTERM', () => {
  console.log('[metrics-agent] Shutting down...');
  server.close(() => process.exit(0));
});

process.on('SIGINT', () => {
  console.log('[metrics-agent] Interrupted, shutting down...');
  server.close(() => process.exit(0));
});
